Feat/Scrum 129 출생 연도 및 프로필 설정 UI

noeyiz opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [x] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경
유저가 자신의 출생 연도, 닉네임, 프로필 사진을 설정할 수 있는 화면이 필요합니다.
<!-- 해당 PR에서 개발한 내용에 대해 알아야 하는 배경지식을 작성해주세요. -->

## 목표
피그마 디자인과 일치하는 프로필 설정 UI를 구현합니다.
<!-- 해당 PR에서 개발/수정 한 기능의 목표를 작성해주세요. -->

## 결과 (Optional)

### 출생 연도 설정

|<img width="552" alt="Screenshot 2024-06-19 at 2 18 21 AM" src="https://github.com/project-lyrics/app-iOS/assets/116897060/6adb1a8f-3044-4d0f-a3aa-5e0c2c4f2c0b">|<img width="552" alt="Screenshot 2024-06-19 at 2 18 18 AM" src="https://github.com/project-lyrics/app-iOS/assets/116897060/f180c613-7ab5-4a44-bc09-99e45275b473">|
|-|-|

### 프로필 설정

|정상 입력|한글, 영어, 숫자 외 입력|
|-|-|
|<img src="https://github.com/project-lyrics/app-iOS/assets/116897060/d1ff16a6-07f4-40e0-aebf-a568ec8c9bc0">|<img src="https://github.com/project-lyrics/app-iOS/assets/116897060/71867361-4ac3-45d7-9b49-6749b0661359">|

|10자 초과|한글, 영어, 숫자 외 입력 + 10자 초과|
|-|-|
|<img src="https://github.com/project-lyrics/app-iOS/assets/116897060/00bd2385-98fe-4c0b-a1b2-f37db0a647d6">|<img src="https://github.com/project-lyrics/app-iOS/assets/116897060/2dd46331-188e-4c2d-88b1-26b471dbd698">|

<img width="250" src="https://github.com/project-lyrics/app-iOS/assets/116897060/5557f2f4-7646-48da-bbc9-678947b584cb">

<!-- 개발한 내용의 결과를 설명해주세요. !-->

noeyiz committed:
feat: NicknameTextView 구현

- 한/영/숫자로 10자 이내의 닉네임을 입력하는 텍스트 필드 뷰 구현
- 조건을 충족하지 않을 시 상황에 맞는 에러 메시지 출력

noeyiz committed:
feat: ProfileView 구현

noeyiz committed:
refactor: UILabel+ setTextWithLineHeight 메서드에서 numberOfLines 설정하도록 변경

noeyiz committed:
feat: EditProfileView 구현

- 제네릭 UIView 타입을 사용하는 BottomSheetViewController 구현
- Shared 모듈에 FlexLayout 의존성 추가
- EditProfileView/ViewController 구현

noeyiz committed:
design: UI 수정

- BottomSheetViewController 애니메이션 시 dimmedView alpha값 변경
- Dim 컬러 alpha값 수정
- 피그마 디자인 여백 변경에 따른 UI 수정

noeyiz committed:
refactor: 상위 모듈만 임포트하도록 변경

noeyiz committed:
refactor: 매직 넘버를 ProfileCharacterType의 개수로 변경

noeyiz committed:
chore: 괄호 누락 수정

noeyiz committed:
design: 피그마 디자인 변경에 따른 UI 수정

noeyiz committed:
feat: SelectBirthYearView 구현

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/EditProfileView.swift:
```diff
@@ -0,0 +1,88 @@
+        return label
+    }()
+    
+    let xButton = {
+        let button = UIButton()
+        button.setImage(FeelinImages.xLight, for: .normal)
+        return button
+    }()
```
@inwoodev  @noeyiz 
이전 인우님 PR에서 self로 접근해서 만들어지는 버튼이 액션이 있는 경우, lazy var로 하신다고 하셨던가요? 
혹시 xButton은 lazy var일 필요는 없을까요?

noeyiz replied:
저는 해당 프로퍼티가 초기화될 때 다른 프로퍼티나 메서드를 사용하는 경우에만 lazy var를 사용하고, 그 외에는 let을 사용했어요 ! xButton은 외부 요소를 사용하지 않아서 let으로 선언했습니당

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/NicknameTextFieldView.swift:
```diff
@@ -0,0 +1,144 @@
+        label.textColor = Colors.gray02
+        label.textAlignment = .right
+        return label
+    }()
```
lazy var와 UILabel을 메서드해서 사용하신 이유가 있을까요? 
let으로 UILabel을 생성해서 사용해도 코드의 길이는 똑같지 않을까요?

noeyiz replied:
어머 그럴 것 같아요 아마 textColor 속성을 createDescriptionLabel 메서드에 추가했었다가 없애면서 이렇게 남은 것 같습니다 :sweat_smile: 수정해두겠습니당 ~

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/ProfileEditButton.swift:
```diff
@@ -0,0 +1,98 @@
+            .alignItems(.end)
+            .define { flex in
+            flex.addItem(profileImageView)
+                .width(176)
+                .height(176)
+                .marginLeft(20)
```
**P3**

이 크기는 비율로 하는게 어떨까요?
iPhone mini에서는 너무 크지 않을까요?

noeyiz replied:
오 ! 그럴 것 같아요 비율로 수정하겠습니다 ㅎㅎ 👍🏻

derrickkim0109 commented on Projects/Shared/DesignSystem/Sources/Models/ProfileCharacterType.swift:
```diff
@@ -0,0 +1,26 @@
+import UIKit
+
+public enum ProfileCharacterType: CaseIterable {
+    case profileShortHair
+    case profileBraidedHair
+    case profile5To5PartedHair
+    case profilePoopHair
```
**P4**

enum 타입에 Profile이란 네이밍이 명시되어 있다면 case에서는 profile을 빼도되지 않을까요?

noeyiz replied:
**5To5PartedHair** 이 케이스 때문에 profile을 붙이게 되었습니당 (ㅠ ㅠ) 이름이 숫자로 시작하게 되어서요 .. profile을 빼고 fiveTofivePartedHair 또는 partedHair로 수정할까요 ?

derrickkim0109 replied:
저는 그게 좋을 것 같은데 인우님은 어떻게 생각하시나요?

inwoodev replied:
태현님 방향도 좋습니다~

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/NicknameTextFieldView.swift:
```diff
@@ -0,0 +1,144 @@
+        textField.textColor = Colors.gray08
+        textField.borderStyle = .none
+        textField.autocorrectionType = .no
+        textField.autocapitalizationType = .none
```
오 이런 설정이 있군요.ㅎㅎ

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/NicknameTextFieldView.swift:
```diff
@@ -0,0 +1,144 @@
+}
+
+private extension NicknameTextFieldView {
+    @objc private func textFieldEditingChanged(_ textField: UITextField) {
```
extension과 메서드에 중복해서 private을 처리하고 있는 것 같아요. 
108번 라인처럼 extension에 선언해두면 좋을 것 같습니다~!

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/ProfileView.swift:
```diff
@@ -0,0 +1,94 @@
+    
+    let profileEditButton = ProfileEditButton()
+    
+    private lazy var guideLabel = {
+        let label = UILabel()
+        label.setTextWithLineHeight(
+            "*닉네임은 한/영/숫자 상관없이 \(maxNicknameLength)자 이내\n(공백, 특수문자, 이모티콘 사용 불가)",
+            lineHeight: 20
+        )
+        label.font = SharedDesignSystemFontFamily.Pretendard.regular.font(size: 14)
+        label.textColor = Colors.gray04
+        return label
+    }()
```
이 Label도 Lazy var로 선언하신 이유가 있을까요?

noeyiz replied:
ProfileView의 `maxNicknameLength` 프로퍼티를 `guideLabel` 초기화 시 사용하기 때문에 lazy var로 선언했습니다 !
( maxNicknameLength가 먼저 초기화된 이후에 guideLabel이 초기화되도록 보장하기 위해 )

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/ProfileEditButton.swift:
```diff
@@ -0,0 +1,98 @@
+        return view
+    }()
+    
+    private let editImageView = {
+        let imageView = UIImageView()
+        imageView.image = FeelinImages.pencil
+        return imageView
+    }()
```
imageView.contentMode 설정이 빠진 것 같아요. default 값인 scaleToFill로 설정되면 되는 걸까요?

noeyiz replied:
scaleAspectFit로 설정해야 할 것 같습니다 ! 발견해주셔서 감사합니다 ㅎㅎ

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/ProfileCharacterCell.swift:
```diff
@@ -0,0 +1,74 @@
+    
+    // MARK: - components
+    
+    private let imageView = UIImageView()
```
imageView.contentMode 설정이 필요하지 않나요?

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/SelectBirthYearView.swift:
```diff
@@ -0,0 +1,62 @@
+    // MARK: - components
+    private let flexContainer = UIView()
+    
+    let doneButton = {
+        let button = UIButton()
+        button.setTitle("Done", for: .normal)
+        button.setTitleColor(Colors.gray09, for: .normal)
+        return button
+    }()
+    
+    lazy var pickerView = UIPickerView()
```
doneButton, pickerView는 어떤 기준으로 lazy var 선언을 하시는 걸까요?

noeyiz replied:
이것은 저의 실수입니다 .. :face_holding_back_tears: 발견해주셔서 감사해용 let으로 수정하겠습니다 ~

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/ViewControllers/EditProfileViewController.swift:
```diff
@@ -0,0 +1,98 @@
+    @objc private func selectButtonDidTap() {
+        dismiss(animated: true)
+    }
+    
```
13, 43라인에 개행이 있어요~

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/ViewControllers/ProfileViewController.swift:
```diff
@@ -0,0 +1,52 @@
+    }
+    
+    @objc private func profileEditButtonDidTap() {
+        let editProfileViewController = EditProfileViewController(bottomSheetHeight: 284)
+        editProfileViewController.modalPresentationStyle = .overFullScreen
+        present(editProfileViewController, animated: false)
```
@inwoodev , @noeyiz 
Coordinator에서 present도 관리하는 건 어떻게 생각하세요?
NavigationController만 관리하는 걸까요?

**P4**

ViewController 생성과 관련된 부분을 관리하는 관점에서 Coordinator에서 관리하는 건 어떤가요?

noeyiz replied:
저는 좋습니당 👍🏻 아직 온보딩 화면에 네비게이션 로직이 없어서 우선 이렇게 구현해두었는데, 옮기면 좋을 것 같아요 !

inwoodev replied:
Coordinator에서 관리해도 괜찮을 듯 해요~~ 다만 요건 추후 온보딩 비즈니스 로직을 구현할 때 수정해도 무방하지 않나 생각이 들긴합니다.

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/ViewControllers/UserInformationViewController.swift:
```diff
@@ -19,9 +21,28 @@ public final class UserInformationViewController: UIViewController {
+    }
+    
+    @objc private func birthYearDropDownButtonDidTap() {
+        let selectBirthYearViewController = SelectBirthYearViewController(
+            bottomSheetHeight: 284,
+            baseYear: baseBirthYear
+        )
+        selectBirthYearViewController.modalPresentationStyle = .overFullScreen
+        selectBirthYearViewController.delegate = self
+        present(selectBirthYearViewController, animated: false)
```
여기도 같이 Coordinator에 대한 의논이 필요할 것 같아요.

derrickkim0109 commented on Projects/Shared/Util/Sources/Extensions/String+.swift:
```diff
@@ -0,0 +1,16 @@
+import Foundation
+
+extension String {
+    public var containsOnlyAllowedCharacters: Bool {
```
contain에 s가 붙어있어요.

noeyiz replied:
오홍 이건 깊게 생각한 부분은 아니었는데, 지금 생각을 해보자면 ... isEmpty처럼 단수형 동사가 오는 게 일반적이니까 contains가 맞지 않을까 하는 생각이 듭니다 ...

derrickkim0109 replied:
그렇군요~ 좋아요~! 제가 놓친 것 같습니다.

derrickkim0109 reviewed (commented):
수고하셨습니다~!

진짜 빠르시네요 ㅎㅎㅎ

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/Components/NicknameTextFieldView.swift:
```diff
@@ -0,0 +1,145 @@
+//
+//  NicknameTextFieldView.swift
+//  FeatureOnboardingInterface
```
해당 컴포넌트는 추후 닉네임을 수정하는 화면에서도 활용될 것 같아서 SharedDesignSystem쪽으로 옮기는 건 어떨까요?
p3

inwoodev replied:
해당 컴포넌트는 추후 닉네임을 수정하는 화면에서도 활용될 것 같아서 SharedDesignSystem쪽으로 옮기는 건 어떨까요?
p3

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/Components/NicknameTextFieldView.swift:
```diff
@@ -0,0 +1,145 @@
+    
+    // MARK: - init
+    
+    init(maxNicknameLength: Int) {
+        self.maxNicknameLength = maxNicknameLength
+        
+        super.init(frame: .zero)
+        
+        setUpLayout()
+        textField.addTarget(
+            self,
+            action: #selector(textFieldEditingChanged),
+            for: .editingChanged
+        )
+    }
```
일단 현재로서는 닉네임텍스트에 해당 뷰가 사용되긴한데요. 피그마도 보시면 아시겠지만 해당 컴포넌트는 디자인팀 쪽에서는 공용 컴포넌트로 생각하고 만들어 주신것 같아요.
제 생각에는 NickNameTextView도 괜찮지만 조금 더 공용컴포넌트처럼 제작해도 오버엔지니어링은 아니지 않을까요~?
가령 placeholder, regex정도만 초기화시 주입 받는다면 좀 더 활용도가 높아질것 같습니다~
(물론 하단에 색상이 변경되는 messageLabel 까지 고려한다면 머리가 아프긴하네요 ㅎㅎ;
또 드는 생각인데 만약 공용컴포넌트화 한다면 messageLabel도 별도의 뷰로 때어놓고 validation에 따라 색상이 변경되는 로직을 두어도 괜찮을 듯 합니다 :)
)

아 그리고 해당 입력창이 한 줄 이상 넘지 않는다면 textField가 들어가는게 조금 더 목적성 측면에 부합할듯하네요~
p4

noeyiz replied:
피그마 공용 컴포넌트 모아놓은 부분 확인했습니다 ! 👍🏻
한 줄 이상 입력을 받지 않도록 디자인 된 것인지 디자인팀에게 확인한 후 공용 컴포넌트화 해보겠습니다 ~
validation에 따라 색상이 변경되는 label로 분리하는 것도 좋은 것 같아요 ! 감사합니당

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/Components/ProfileEditButton.swift:
```diff
@@ -0,0 +1,98 @@
+final class ProfileEditButton: UIButton {
+    // MARK: - components
+    
+    private let flexContainer: UIView = {
+        let view = UIView()
+        view.isUserInteractionEnabled = false
+        return view
+    }()
```
전 왜 클로저로 생성하면서 `isUserInteractionEnabled` 설정하는걸 생각 못했을까요. 이렇게 하니 로직이 하나로 묶여서 더 좋네요 👍

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/Components/ProfileEditButton.swift:
```diff
@@ -0,0 +1,98 @@
+                        .height(24)
+                        .alignSelf(.center)
+                }
+        }
```
아 edit button 이미지를 하나로 만들어드리도록 디자인팀에 부탁드릴걸 그랬네요..ㅠㅠ 고생하셨습니다!

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/Components/ProfileView.swift:
```diff
@@ -0,0 +1,93 @@
+import SharedDesignSystem
+
+import FlexLayout
+import PinLayout
```
@noeyiz @derrickkim0109 

저 궁금한게 swift framework와 그 외 모듈 import 부분에 개행을 두는 것 뿐 아니라
외부라이브러리 사이에도 개행을 두는걸까요?

noeyiz replied:
스타일 셰어 가이드에는 직접 만든 모듈? 라이브러리? 에 대한 이야기는 없어서 저는 아래와 같은 기준으로 개행을 두고 있었어요 !

```
1. 내장 프레임워크

2. 만든 라이브러리(?)

3. 외부 라이브러리
```

inwoodev commented on Projects/Shared/Util/Sources/Extensions/String+.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+extension String {
+    public var containsOnlyAllowedCharacters: Bool {
+        // 한글, 영문, 숫자만 포함되는 정규식
+        let regex = "^[ㄱ-ㅎ가-힣a-zA-Z0-9]*$"
+        return NSPredicate(format: "SELF MATCHES %@", regex).evaluate(with: self)
+    }
+}
```
p5 입니다.

이건 지극히 제 개인적인 생각인데 String extension에 위 기능을 추가할 경우 String을 사용하는 모든 경우에서 해당 변수가 보이는게 맞나 하는 의문이 사알짝 들긴했습니다.

noeyiz replied:
오홍 생각해보니 해당 정규식은 닉네임 설정할 때만 사용할 것 같으니 말씀해주신 것처럼 모든 경우에서 보일 필요가 없을 것 같아요 ! 수정하도록 하겠습니다 ㅎㅎ

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/Components/EditProfileView.swift:
```diff
@@ -0,0 +1,99 @@
+    private let profileCharacterCollectionView = ProfileCharacterCollectionView()
+    
+    lazy var selectButton = FeelinConfirmButton(initialEnabled: true, title: "선택")
+    
```
요건 lazy가 아니어도 될듯 해요~ 그리고 parameter 두개니까...ㅎㅎ @derrickkim0109  맞나요 태현님??? ㅋㅋㅋㅋ

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/Components/ProfileCharacterCell.swift:
```diff
@@ -0,0 +1,75 @@
+    func setSelected(_ isSelected: Bool) {
+        layer.borderColor = isSelected ? Colors.active.cgColor : Colors.disabled.cgColor
+    }
+}
```
lgtm 👍

inwoodev commented on Projects/Shared/DesignSystem/Sources/BottomSheetViewController.swift:
```diff
@@ -0,0 +1,124 @@
+            }
+        }
+    }
+}
```
장인의 손길이 느껴집니다. 하나 더 배워갑니다~ 고생하셨습니다 👍 👍

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/Components/ProfileCharacterCollectionView.swift:
```diff
@@ -33,7 +33,8 @@ final class ProfileCharacterCollectionView: UICollectionView {
         
-        let cellWidth = (frame.width - flowLayout.minimumInteritemSpacing * 3) / 4
+        let numOfItems = ProfileCharacterType.allCases.count
+        let cellWidth = (frame.width - flowLayout.minimumInteritemSpacing * (numOfItems - 1) / numOfItems
```
요런 디테일한 부분 수정 👍 👍

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/ViewControllers/SelectBirthYearViewController.swift:
```diff
@@ -0,0 +1,96 @@
+    func setBitrhYear(year: Int)
+}
+
+public final class SelectBirthYearViewController: BottomSheetViewController<SelectBirthYearView> {
```
요런 상속 구조 너무 좋습니다~ 지연님이 만들어 주신 BottomSheetViewController잘 사용 해 볼게요~

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/ViewControllers/SelectBirthYearViewController.swift:
```diff
@@ -0,0 +1,96 @@
+    
+    // MARK: - init
+    
+    public init(bottomSheetHeight: CGFloat, baseYear: Int) {
+        self.selectedYear = baseYear
+        
+        super.init(bottomSheetHeight: bottomSheetHeight)
+        
+        setUpPickerView(baseYear: baseYear)
+        setUpAction()
+    }
```
@derrickkim0109 소환~~ ㅋㅋㅋㅋ

derrickkim0109 replied:
이건 어떤 부분말하시는 걸까요?

inwoodev replied:
아..아닙니다. 취소입니다 :)

inwoodev reviewed (approved):
고생하셨어요~ 지연님!! 체리픽 하기 번거로우셨을 것 같은데 그래도 의견 받아주셔서 감사합니다~

몇 자 코멘트 남겨두었습니다. 마이너한 부분이라 생각되어서 일단 Approve하겠습니다.

필요하다 생각하시는 부분이 있다면 수정하시고 머지하셔도 좋을 듯 합니다 :)

derrickkim0109 reviewed (approved):
수고하셨습니다~

noeyiz committed:
refactor: 컨벤션, 불필요한 lazy var, extension, 접근 지정자 수정

noeyiz committed:
remove: UserInterfaceStyle 라이트 모드 고정으로 설정하는 코드 제거

noeyiz committed:
design: 프로필 버튼 고정 크기가 아닌 비율로 변경

noeyiz committed:
feat: NicknameTextField -> FeelinLineInputField로 변경

- 유효성에 따라 텍스트 컬러가 바뀌는 FeelinAlertLabel 추가
- 기존 NicknameTextField를 공용 컴포넌트화 한 FeelinLineInputField 추가

noeyiz commented on Projects/Shared/DesignSystem/Sources/Views/FeelinAlertLabel.swift:
```diff
@@ -0,0 +1,28 @@
+
+import UIKit
+
+public final class FeelinAlertLabel: UILabel {
+    /// `isValid`가 true이면 텍스트 색상을 `Colors.alertSuccess`로 설정하고,
+    /// false이면 `Colors.alertWarning`으로 설정합니다.
+    public var isValid: Bool = true {
+        didSet {
+            textColor = isValid ? Colors.alertSuccess : Colors.alertWarning
+        }
+    }
+    
+    public override init(frame: CGRect) {
+        super.init(frame: frame)
+        isValid = true
+    }
+    
+    @available(*, unavailable)
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+}
```
유효성에 따라 텍스트 컬러가 바뀌는 UILabel ... 제가 이해한 내용이 맞을까요 ? 구현하고 보니 텍스트 컬러만 바꿔주는 거라 따로 빼는 게 유의미한가 싶은 생각이 들어서 ... 제가 이해를 잘못한 것일 수도 있으니 편하게 의견 주세욥 흑흑 ... 🥹

derrickkim0109 replied:
음.. 유효성에 따라 텍스트 컬러만 바꿔주는 거면 UILabel에서 extension으로 빼서 설정해줘도 되지 않을까요?

inwoodev replied:
이미 남기긴했지만 FeelinLineIntpuField와 NickNameTextFieldView를 나눠서 구현할 때 `NickNameTextFieldView`에 일부로 포함시키면 이렇게 별도로 빼지 않아도 될것 같다는 생각이 듭니다~~ :)

noeyiz committed:
refactor(FeelinLineInputField): 유효성 검사 여러 개 설정할 수 있도록 변경

noeyiz commented on Projects/Feature/Onboarding/Interface/Sources/Components/ProfileView.swift:
```diff
@@ -0,0 +1,105 @@
+        placeholder: "텍스트",
+        lengthExceededMessage: "1~10자의 닉네임을 사용해주세요",
+        validationRules: [
+            { (nickname) in
+                let regex = "^[ㄱ-ㅎ가-힣a-zA-Z0-9]*$"
+                let isValid = NSPredicate(format: "SELF MATCHES %@", regex).evaluate(with: nickname)
+                return (isValid, isValid ? nil : "공백, 특수문자, 이모티콘은 사용 불가합니다")
+            }
```
이런 식으로 유효성 검사 결과 및 메시지를 리턴하는 클로저의 배열로 받도록 했는데 ... 더 좋은 방법이 있을지 함께 고민해주시면 감사하겠습니다 !

derrickkim0109 replied:
이 방식 좋은 것 같아요~!

inwoodev replied:
어제 말씀드렸다시피 저는 개인적으로 `validationRules`를 클로저로 받을 필요는 없을것 같다는 생각이 들었어요. 클로저 또한 (String) -> (Bool, String)인 부분도 조금 모호하다고 생각이 들었구요.
 
 음 이렇게 가면 어떨까요?
 공용 컴포넌트는 FeelinLineIntpuField로 두고 (텍스트필드와 line 정도)
 이걸 활용해서 NickNameTextFieldView를 구현하는거죠! (텍스트필드, 하단 label & 글자 카운트 포함)
 
 그리고 정규식 관련해서는 요 [깃헙](https://github.com/sindresorhus/Regex/blob/main/Sources/Regex/Regex.swift) 도 한 번 참고 해 보시면 좋을듯합니다~

noeyiz replied:
같이 고민해주셔서 감사합니다 ! 저도 객체 생성할 때 클로저로 유효성 검사 로직 및 사용될 메시지 등을 넘겨주는 게 마음에 들지 않아서 코멘트를 남겨놓은 것이었는데 🥲 .. 피그마 디자인 상으로 라인 인풋 필드에 해당하는 게 현재 글자수와 메시지까지가 포함이라 만약 NicknameTextFieldView를 따로 구현하게 된다면 나중에 다른 로직을 검사하는 라인 인풋 필드가 필요함에따라 로직만 다른 객체가 계속 늘어날 수 있을 것 같다는 생각이 들어서요 ㅠ _ ㅠ 그러면 UI만 공용 컴포넌트로 구현해놓고, 유효성 검사 로직은 delegate 패턴으로 외부에서 처리하게 하는 방법은 혹시 어떨까요 ..?

noeyiz reviewed (commented):
기존 NicknameTextFieldView를 공용 컴포넌트화하여 FeelinLineInputField로 변경했습니다. 텍스트 길이 외 유효성 검사 로직이 여러 개 정의될 수 있을 것 같아서 고민하다가 (유효성 검사 결과, 메시지)를 리턴하는 클로저의 배열을 인자로 받도록 했는데 좋은 방법인지 잘 모르겠습니다 ... 리뷰 브탁드립니다 🙇🏻‍♀️

derrickkim0109 reviewed (approved):
수고하셨습니다~!

noeyiz committed:
refactor: FeelinLineInputField 수정

- 텍스트필드와 구분선만 가지도록 수정

noeyiz committed:
feat: ProfileView에 닉네임 유효성 검증 로직 추가

noeyiz committed:
feat: 출생 연도 범위 수정

- 만 14~100세로 출생 연도 범위 수정

noeyiz commented:
- FeelinLineInputField를 텍스트 필드와 구분선만을 가진 컴포넌트로 수정했습니다.
- 유효성 검증과 관련된 UI(경고 메시지와 현재 글자 수 표시)와 로직은 ProfielView에서 관리하도록 수정했습니다.
- 지난 주 화요일에 있던 회의에 따라 출생 연도의 범위를 만 14~100세로 변경했습니다.

감사합니다..!
